### ISSUE TITLE
[FEATURE] Implement unified processing for LAMBDAS and call-by-name for both built-in and user-defined operators

### ISSUE THREAD

**Issue by Kukovec** [informalsystems/apalache#872]
Since #866 exposes Apalache to call-by-name, and likely an influx in LAMBDA usage, we should come up with a unified way of representing all of these cases. Consider the following expressions:
```
LET A(p,q) == Append(p,q) IN FoldSeq( A, v, seq ) \* (1)

FoldSeq( Append, v, seq ) \* (2)

FoldSeq( LAMBDA p,q : Append(p,q), v, seq ) \* (3)
```
Currently, Apalache would fail on (2) and (3), since all of the above expressions either have different internal representations, or are incompatible with typechecking (e.g. in (2), type annotations cannot be supplied, but they should also not be necessary, as the types of built-in operators are known\*).

We should preprocess all of the above expressions into the following normal form:
```
FoldSeq( X, v, seq ) ---> FoldSeq( LET LocalX(p,q) == X(p,q) IN LocalX , v, seq )
```

\* possibly polymorphic/polyadic!

**Comment by Kukovec** [comment]
After some investigation, it does not seem likely that we will be supporting LAMBDAs or built-in-operators in the near future, mainly due to the constraints imposed by requiring type annotations (there's currently no way to annotate a LAMBDA).

**Comment by svalaskevicius** [comment]
interesting, I'm new to apalache and may not understand how it works yet, but I was able to annotate a lambda and pass a typecheck with ` (* @type: MESSAGE => Bool; *) LAMBDA ...`..

only to get an error in checking `unexpected expression: undeclared operator LAMBDA$2` :)

**Comment by Kukovec** [comment]
Yes, while this information is present at typechecking, it (as well as other bits, which cause the error you're getting) gets erased in an internal conversion step later on, because several passes (that mostly predate typechecking) were designed under the assumption that we wouldn't be supporting lambdas, so the support we do have for lambdas at the moment is often ad hoc. We need a better internal representation of lambdas before we can support them properly.
For example, the below spec passes the first typecheck, but fails the 2nd: `check --inv=Inv --length=1 test.tla`
```tla
---------- MODULE test ----------
EXTENDS Apalache

Init == TRUE
Next == TRUE
Inv == FoldSeq( (* @type: (Int,Int) => Int; *) LAMBDA p,q: p, 0, <<>> ) = 0

====================
``` 
```
...
PASS #1: TypeCheckerSnowcat                                       I@16:04:55.242
 > Running Snowcat .::.                                           I@16:04:55.242
 > Your types are great!                                          I@16:04:55.335
 > All expressions are typed 
...
PASS #12: PostTypeCheckerSnowcat                                  I@16:04:55.557
 > Running Snowcat .::.                                           I@16:04:55.557
[test.tla:6:48-6:60]: No annotation found for LAMBDA$1. Make sure that you've put one in front of LAMBDA$1. E@16:04:55.563
```
Until this is addressed, I'd recommend writing LET-IN instead of lambdas, e.g. 
```tla
LET \* @type: (T1,T2) => T1;
    NotLambda(p,q) == ... 
IN FoldSeq( NotLambda , v, seq )
```

**Comment by svalaskevicius** [comment]
Right, thanks for the info, and the workaround! :)

**Comment by konnov** [comment]
@Kukovec is it still an issue?

**Comment by Kukovec** [comment]
I'm not sure we have consensus on whether or not we want to support lambdas at all in the future, but the annotation-before-lambda parsing hasn't been fixed yet.

**Comment by Kukovec** [comment]
Rather, I think we should use this issue to keep track of what the lambda form we end up supporting is going to be. Unless you want to make a whole new issue for that?

**Comment by Kukovec** [comment]
I guess you can close this and keep https://github.com/informalsystems/apalache/issues/2532
